Add support for SFNode editing in Web PROTO

ad-daniel opened the pull request:
**Description**

Use-case:
A user wants to create a library of custom objects as part of the process of creating a competition. In the competition, participants have to assemble their own robot based on a limited library of components to choose from. The programming of the controller can already be achieved without requiring the installation of webots, however the configuration of the robot/object cannot.

This PR allows the customization of PROTO files (specifically SFNodes) on the web without requiring the installation of Webots.

https://proto.webots.cloud/run?version=proto&url=https://github.com/BenjaminDeleze/sandbox/blob/master/protoExample/protos/OfficeChair.proto

- [x] list SFNode parameters in the interface
- [x] add node selector window
- [x] add compatibility filter mechanism (slot type, robot ancestors, ...)
- [x] handle insertion in node structure
   - [x] add library mechanism to choose PROTO from
   - [x] generate and/or load meta-data (from proto-list.xml) 
   - [x] handle insertion in the node-structure (and js notification)
   - [x] propagate actions (insertion and deletions) to DEF-USE

BenjaminDeleze commented on resources/web/wwi/FloatingProtoParameterWindow.js:
```diff
@@ -515,17 +535,20 @@ export default class FloatingProtoParameterWindow extends FloatingWindow {
 
   #rotationOnChange(node) {
-    const object = {'x': node.inputs[0].value, 'y': node.inputs[1].value, 'z': node.inputs[2].value, 'a': node.inputs[3].value};
+    const object = {'x': parseFloat(node.inputs[0].value), 'y': parseFloat(node.inputs[1].value),
+      'z': parseFloat(node.inputs[2].value), 'a': parseFloat(node.inputs[3].value)};
```
I got this warning: 
` Object properties must go on a new line if they aren't all on the same line. (object-property-newline)`

BenjaminDeleze commented on resources/web/wwi/NodeSelectorWindow.js:
```diff
@@ -0,0 +1,384 @@
+    return new Promise((resolve, reject) => {
+      const xmlhttp = new XMLHttpRequest();
+      xmlhttp.open('GET', 'https://cyberbotics.com/wwi/proto/protoVisualizer/temporary-proto-list.xml', true);
+      xmlhttp.onreadystatechange = async () => {
```
```suggestion
      xmlhttp.onreadystatechange = async() => {
```

BenjaminDeleze commented on resources/web/wwi/NodeSelectorWindow.js:
```diff
@@ -0,0 +1,384 @@
+    nodeFilter.id = 'node-filter';
+    nodeFilter.className = 'node-filter';
+    const p = document.createElement('p');
+    p.style.margin = 'auto 10px auto 0px'
```
```suggestion
    p.style.margin = 'auto 10px auto 0px';
```

BenjaminDeleze commented on resources/web/wwi/NodeSelectorWindow.js:
```diff
@@ -0,0 +1,384 @@
+      if (this.parameter.value.value === null || !compatibleNodes.includes(this.parameter.value.value.name))
+        this.selection = ol.children[0]; // select the first item if the parameter doesn't currently contain anything
+      else
+        this.selection = ol.children[compatibleNodes.indexOf(this.parameter.value.value.name)]
```
```suggestion
        this.selection = ol.children[compatibleNodes.indexOf(this.parameter.value.value.name)];
```

BenjaminDeleze commented on resources/web/wwi/NodeSelectorWindow.js:
```diff
@@ -0,0 +1,384 @@
+        this.selection = ol.children[compatibleNodes.indexOf(this.parameter.value.value.name)]
+
+      this.selection.style.backgroundColor = '#007acc';
+
```
```suggestion
```

BenjaminDeleze commented on resources/web/wwi/NodeSelectorWindow.js:
```diff
@@ -0,0 +1,384 @@
+      this.populateNodeInfo(this.selection.innerText);
+    };
+
+    button.ondblclick = async (item) => this.insertNode(item.target.innerText);
```
```suggestion
    button.ondblclick = async(item) => this.insertNode(item.target.innerText);
```

BenjaminDeleze commented on resources/web/wwi/NodeSelectorWindow.js:
```diff
@@ -0,0 +1,384 @@
+      const url = info.url;
+      nodeImage.src = url.slice(0, url.lastIndexOf('/') + 1) + 'icons/' + protoName + '.png';
+      description.innerHTML = info.description;
+      license.innerHTML = 'License:&nbsp;<i>' + info.license + '</i>'
```
```suggestion
      license.innerHTML = 'License:&nbsp;<i>' + info.license + '</i>';
```

BenjaminDeleze commented on resources/web/wwi/NodeSelectorWindow.js:
```diff
@@ -0,0 +1,384 @@
+
+  isAllowedToInsert(baseType, slotType) {
+    if (typeof this.parameter === 'undefined')
+      throw new Error('The parameter is expected to be defined prior to checking node compatibility.')
```
```suggestion
      throw new Error('The parameter is expected to be defined prior to checking node compatibility.');
```

BenjaminDeleze commented on resources/web/wwi/NodeSelectorWindow.js:
```diff
@@ -0,0 +1,384 @@
+    filterInput.value = '';
+
+    if (!(parameter instanceof Parameter))
+      throw new Error('Cannot display node selector unless a parameter is provided.')
```
```suggestion
      throw new Error('Cannot display node selector unless a parameter is provided.');
```

BenjaminDeleze commented on resources/web/wwi/Parser.js:
```diff
@@ -433,11 +449,22 @@ export default class Parser {
+    const viewpoint = new WbViewpoint(id, fieldOfView, orientation, position, exposure, bloomThreshold, near, far,
+      followSmoothness, followedId, ambientOcclusionRadius);
+
+    if (typeof parentNode !== 'undefined') {
+      if (parentNode instanceof WbGroup)
+        parentNode.children.push(viewpoint);
```
I think that `parentNode`  is not defined

BenjaminDeleze commented on resources/web/wwi/ProtoManager.js:
```diff
@@ -1,6 +1,7 @@
 import Node from './protoVisualizer/Node.js';
 import { MFNode, SFNode, stringifyType } from './protoVisualizer/Vrml.js';
 import {getAnId} from './nodes/utils/id_provider.js';
+import WbWorld from './nodes/WbWorld.js';
```
```suggestion
```

BenjaminDeleze commented on resources/web/wwi/WebotsView.js:
```diff
@@ -66,7 +66,7 @@ export default class WebotsView extends HTMLElement {
 
   #loadScript(scriptUrl) {
-    return new Promise(function(resolve, reject) {
+    return new Promise(function (resolve, reject) {
```
```suggestion
    return new Promise(function(resolve, reject) {
```

BenjaminDeleze commented on resources/web/wwi/X3dScene.js:
```diff
@@ -179,21 +178,25 @@ export default class X3dScene {
-    if (typeof parentId !== 'undefined' && parentId > 0) {
+    if (typeof parentId !== 'undefined') {
       parentNode = WbWorld.instance.nodes.get('n' + parentId);
-      const ancestor = getAncestor(parentNode);
```
we can remove the `getAncestor` import

BenjaminDeleze commented on resources/web/wwi/X3dScene.js:
```diff
@@ -165,11 +164,11 @@ export default class X3dScene {
     xmlhttp.open('GET', url, true);
     xmlhttp.overrideMimeType('plain/text');
-    xmlhttp.onreadystatechange = () => {
+    xmlhttp.onreadystatechange = async () => {
```
```suggestion
    xmlhttp.onreadystatechange = async() => {
```

BenjaminDeleze commented on resources/web/wwi/X3dScene.js:
```diff
@@ -179,21 +178,25 @@ export default class X3dScene {
+      if (parentNode instanceof WbShape) // TODO: this might be improved with a onchange trigger
+        parentNode.updateAppearance();
+    } else
+      node.finalize()
```
```suggestion
      node.finalize();
```

BenjaminDeleze commented on resources/web/wwi/nodes/WbPbrAppearance.js:
```diff
@@ -215,6 +215,9 @@ export default class WbPbrAppearance extends WbAbstractAppearance {
 
   clone(customID) {
+    if (typeof customID === 'undefined')
+      customID = getAnId()
```
```suggestion
      customID = getAnId();
```

BenjaminDeleze commented on resources/web/wwi/nodes/WbShape.js:
```diff
@@ -18,10 +19,33 @@ export default class WbShape extends WbBaseNode {
+      }
+    }
+  }
+
```
```suggestion
```

BenjaminDeleze commented on resources/web/wwi/protoVisualizer/Node.js:
```diff
@@ -311,9 +314,26 @@ export default class Node {
-    // TODO
+    this.def = new Map();
+
+    for (const [parameterName, parameter] of this.parameters)
```
```suggestion
    for (const [, parameter] of this.parameters)
```
Is this working? that is what my linter is suggesting

BenjaminDeleze commented on resources/web/wwi/protoVisualizer/Parameter.js:
```diff
@@ -70,34 +70,36 @@ export default class Parameter {
-
-      // delete existing node (must be done prior to regeneration or the information is lost)
-      const id = baseNode.id;
+      const id = this.node.getBaseNode().id; // delete existing node (must be done prior to regeneration or the information is lost)
```
Line too long

BenjaminDeleze commented on resources/web/wwi/protoVisualizer/Vrml.js:
```diff
@@ -291,13 +291,13 @@ export class SFRotation extends SingleValue {
+    if (v.x === NaN)
       v.x = 0;
-    if (v.y === '')
+    if (v.y === NaN)
       v.y = 0;
-    if (v.z === '')
+    if (v.z === NaN)
       v.z = 0;
-    if (v.a === '')
+    if (v.a === NaN)
```
We should use the `isNaN()` function to compare with NaN

BenjaminDeleze reviewed (changes requested):
Works well in general, just some remarks:

- When `no nodes match the criteria`, the text is stuck against one side, it would be nicer if it had some space.
- In the configuration window of SFNode, it would nicer if the left margin was bigger.
- Not sure if related with this PR, but when changing the string of some PBRAppearances, lets say I put `top` instead of `back` for the `cardboard`  type, an error is thrown: 
```Requested deletion of node that does not exist```
- Personally, I would inverse the order of the icons: first the configure and then the delete.
- The state of the reset buttons is not saved when moving back and forth in protos
- The download name is not saved when going in subproto
- When trying with the Nao, it finds nothing to suggest for the `jersey` field whereas the `NaoJersey` should be suggested.
- This PR breaks the streaming viewer, the animations and the documentation.
- Not sure if introduced in this PR but it seems that the objects are not always correctly cleaned:
  - go in `https://proto.webots.cloud/run?version=proto&url=https://github.com/BenjaminDeleze/sandbox/blob/master/protoExample/protos/OfficeChair.proto`
  - change enable physics to FALSE
  - Click on webots.cloud to go back to the main menu
  - Load the nao proto
  - Both the nao and the office chair are in the scene

ad-daniel commented:
I believe all points have been addressed, I've uploaded the files to the server

BenjaminDeleze commented on resources/web/wwi/FloatingProtoParameterWindow.js:
```diff
@@ -158,36 +200,47 @@ export default class FloatingProtoParameterWindow extends FloatingWindow {
     buttonContainer.style.gridColumn = '4 / 4';
 
+    if (typeof this.exportName === 'undefined')
+      this.exportName = this.#protoManager.proto.name
+
     const input = document.createElement('input');
     input.type = 'text';
     input.title = 'New proto name';
-    input.value = 'My' + this.#protoManager.proto.name;
+    input.value = this.exportName;
+    input.onchange = (e) => this.exportName = e.target.value;
```
I saw that you removed the `My` prefix.
I added it to avoid conflicts with the base proto and also to show that the user can modify the name

BenjaminDeleze reviewed (changes requested):
Good job! most of the issues are indeed fixed, but I could still find some:

- TextureTransform of the proto parameters has its `resetButton` enable and it throws an error when clicking on the reset button
- When streaming and adding an object to the scene, nothing appears and I get to following error: ```parentNode is undefined at x3dScene.js:182```
  -  I get the same error and the object disappear when modifying a regenerating parameter
  -   Same error when reseting the `spot.wbt` world 
- I experimented a bit with the Tinkerbots (I added it on proto.webots.cloud), and even if I am 2 or 3 levels down in subnodes, the `back` button bring me back to the root? Is it the wanted behavior?
- When adding several levels of subprotos, only the first levels (so parameters of the root) are added as externProto, other are not.

ad-daniel commented:
> TextureTransform of the proto parameters has its resetButton enabled

It's the correct behavior, it's flagged as modified in webots too. The reason is that OfficeChair defines the field as `seatAppearance  Leather { textureTransform TextureTransform { scale 10 10 } }`, so the the `textureTransform` field is non-default with respect to Leather. I've fixed the behavior when clicking the reset button, which was indeed wrong.

> I experimented a bit with the Tinkerbots (I added it on proto.webots.cloud), and even if I am 2 or 3 levels down in subnodes, the back button bring me back to the root?

This was the intended behavior as when discussing with Olivier we agreed this was good enough for now, we care more about the root after all. In the case of the tinkerbots however the layers can get big fairly quickly, so I've changed it and now it returns back up one level instead. We'll need it later one either way, so might as well

BenjaminDeleze commented on resources/web/wwi/X3dScene.js:
```diff
@@ -528,20 +527,19 @@ export default class X3dScene {
       const parentId = data.split(':')[0];
       data = data.substring(data.indexOf(':') + 1);
-      this.loadObject(data, parentId);
+      this.loadObject(data, parentId === "0" ? WbWorld.instance.root.id.replace('n', '') : parentId);
```
```suggestion
      this.loadObject(data, parentId === '0' ? WbWorld.instance.root.id.replace('n', '') : parentId);
```

BenjaminDeleze reviewed (changes requested):
I observed the two following bugs in streaming, however, they are also present in Master so I think we should fix them later and not in this PR:
  - When adding a jersey to a nao it does not work.
  - I the regeneration of the nao is triggered, the updates do not work anymore.

ad-daniel commented:
should probably open an issue about them or they might be forgotten

BenjaminDeleze reviewed (approved):
Good job!
